From d8fe2e9dbf6070440e820e3b6b82242a34b4fffe Mon Sep 17 00:00:00 2001
From: Hugo Peixoto <hugo.peixoto@gmail.com>
Date: Sun, 6 Mar 2011 10:28:46 +0000
Subject: [PATCH] Fixes rails bug #6058.

Propagates 'where' clauses when subquerying is triggered on the UpdateManager.
---
 lib/arel/visitors/to_sql.rb |    1 +
 test/test_select_manager.rb |   14 ++++++++++++++
 2 files changed, 15 insertions(+), 0 deletions(-)

diff --git a/lib/arel/visitors/to_sql.rb b/lib/arel/visitors/to_sql.rb
index f76c149..035d906 100644
--- a/lib/arel/visitors/to_sql.rb
+++ b/lib/arel/visitors/to_sql.rb
@@ -35,6 +35,7 @@ module Arel
         stmt             = Nodes::SelectStatement.new
         core             = stmt.cores.first
         core.froms       = o.relation
+        core.wheres      = o.wheres
         core.projections = [key]
         stmt.limit       = o.limit
         stmt.orders      = o.orders
diff --git a/test/test_select_manager.rb b/test/test_select_manager.rb
index 2fe43aa..410fe38 100644
--- a/test/test_select_manager.rb
+++ b/test/test_select_manager.rb
@@ -713,6 +713,20 @@ module Arel
         }
       end
 
+      it 'copies where clauses when nesting is triggered' do
+        engine  = EngineProxy.new Table.engine
+        table   = Table.new :users
+        manager = Arel::SelectManager.new engine
+        manager.where table[:foo].eq 10
+        manager.take 42
+        manager.from table
+        stmt = manager.compile_update(table[:id] => 1)
+
+        stmt.to_sql.must_be_like %{
+          UPDATE "users" SET "id" = 1 WHERE "users"."id" IN (SELECT "users"."id" FROM "users" WHERE "users"."foo" = 10 LIMIT 42)
+        }
+      end
+
       it 'executes an update statement' do
         engine  = EngineProxy.new Table.engine
         table   = Table.new :users
-- 
1.7.1

